### PR TITLE
feat(calling): line class created

### PR DESCRIPTION
**Pull request by meta-dipanshu-sharma**
# COMPLETES #< [SPARK-422965](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-422965) >

## This pull request addresses

Implementation of Line class and implementing all line properties and methods defined in Jira ticket and [Public calling sdk confluence](https://confluence-eng-gpk2.cisco.com/conf/display/WSDK/Public+Calling+SDK+-+Architecture)

## by making the following changes

Created Line class and declared all the properties and methods.

NOTE: Unit tests will be covered in a separate PR.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [ ] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 1]
```diff
@@ -585,6 +234,43 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
+      // dummy values
+      this.webex.internal.device.userId,
+      this.webex.internal.device.url,
+      lineId,
+      '131212345',
+      '212343435',
+      LineStatus.INACTIVE,
+      ['add1', 'add2'],
+      '321232343',
+      '00',
```
Comment (arun3528) [inline packages/calling/calling/src/CallingClient/CallingClient.ts:254]: what are these numbers here?

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 2]
```diff
@@ -518,6 +163,10 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
   public makeCall = (dest: CallDetails): ICall | undefined => {
     let call;
 
+    // this is a temporary logic to get registration obj
+    // it will change once we have proper lineId and multiple lines as well
+    const {registration} = Object.values(this.lineDict)[0];
```
Comment (arun3528) [inline packages/calling/calling/src/CallingClient/CallingClient.ts:168]: this will change once you move to map

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+    this.clientDeviceUri = clientDeviceUri;
+    this.lineId = lineId;
+    this.phoneNumber = phoneNumber;
+    this.extension = extension;
+    this.status = status;
+    this.sipAddresses = sipAddresses;
+    this.voicemail = voicemail;
+    this.lastSeen = lastSeen;
+    this.keepaliveInterval = keepaliveInterval;
+    this.callKeepaliveInterval = callKeepaliveInterval;
```
Comment (arun3528) [inline packages/calling/calling/src/CallingClient/line/index.ts:119]: Most of them wil be default value you don't have to pass them in , We can have a config file instead

**Comment by sreenara** [comment]
> this should be a map 

Let's use the `Record` type

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 4]
```diff
@@ -225,191 +100,6 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
     });
   }
 
-  /**
-   * Fetches countryCode and region of the client.
-   */
-  private async getClientRegionInfo(): Promise<RegionInfo> {
```
Comment (sreenara) [inline packages/calling/calling/src/CallingClient/CallingClient.ts:231]: AFAIK, this needs to stay in CallingClient because it'll be common for all lines

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 5]
```diff
@@ -225,191 +100,6 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
-      regionInfo.countryCode = '';
-    }
-
-    return regionInfo;
-  }
-
-  /**
-   * Local method for finding the mobius servers.
-   */
-  private async getMobiusServers() {
```
Comment (sreenara) [inline packages/calling/calling/src/CallingClient/CallingClient.ts:284]: same comment for this

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 6]
```diff
@@ -585,6 +234,44 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
+      10,
+      this.callManager,
+      this.mutex,
+      this.sdkConfig
+    );
+
+    this.lineDict[lineId] = lineInstance;
+  }
+
+  /**
```
Comment (sreenara) [inline packages/calling/calling/src/CallingClient/CallingClient.ts:269]: Please add a description here.

**Review by sreenara — CHANGES_REQUESTED** [review]
Please see the comments

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 1 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:254]: These are the temporary dummy values for testing the creation of line object. We won't need it once we get config file from DMS.

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 7]
```diff
@@ -62,22 +41,16 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
 
   private sdkConfig: CallingClientConfig | undefined;
 
-  private primaryMobiusUris: string[];
-
-  private backupMobiusUris: string[];
-
   public mediaEngine: typeof Media;
 
+  private lineDict: object = {};
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:52]: Do we want lineDict to be map`(new Map<string, ILine>())` or object with type `Record<string, ILine>`?
@arun3528 @sreenara

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 2 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:168]: yeah this is a temporary hack as we don't have proper lineId through which we can fetch it.

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 3 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts:119]: I'm not sure if we need to set them as default value or will it come through config file from DMS. @Kesari3008 can suggest.

File: .eslintrc.js [hunk 8]
```diff
@@ -46,34 +46,6 @@ module.exports = {
         '@typescript-eslint/no-explicit-any': 'off',
         'lines-between-class-members': ['error', 'always', {exceptAfterSingleLine: true}],
         'tsdoc/syntax': 'off', // Todo: remove this before merging pr
-        'valid-jsdoc': [
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply .eslintrc.js:49]: As discussed previously as well and suggested from Marcin, we don't want to add jsdoc in ts files so that's why removed this eslint check for jsdoc in ts files. Will add a commit which will remove all existing jsdocs also from calling folder.

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 6 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:269]: as said above, we won't be adding jsdocs from now onwards. We will just have a brief description of what the function is doing. Will add it.

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 5 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:284]: This method is used only inside `register()` method which is now present in `Line` class. Some of the common variables are used in both the methods like `primaryMobiusUris` and `backupMobiusUris`. So we have to keep both these in the same file and as `register()` is part of `Line` class so that's why I have moved this method(`getMobiusServers()`) and `getClientRegionInfo()`(which is called from `getMobiusServers()`) to Line.

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 3 above)
Comment (sreenara) [reply packages/calling/calling/src/CallingClient/line/index.ts:119]: @meta-dipanshu-sharma as of now we aren't going to contact DMS to get the device configuration. So we don't need to keep dummy data. When the Mobius starts supporting multiple lines (which will take some time), we can add this back

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 1 above)
Comment (sreenara) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:254]: @meta-dipanshu-sharma as of now we aren't going to contact DMS to get the device configuration. So we don't need to keep dummy data. When the Mobius starts supporting multiple lines (which will take some time), we can add this back

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 5 above)
Comment (sreenara) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:284]: @meta-dipanshu-sharma can we move these functions out of the register() method and instead keep it in CallingClient itself? Then pass the primary and secondary URIs to the Line while initializing it?

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 4 above)
Comment (Kesari3008) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:231]: That's correct @sreenara. We need to keep these methods in callingClient only. I understand that getMobiusServers is being invoked inside register() API and hence these needed moving due to dependency.

@meta-dipanshu-sharma  Can we try an approach where getMobiusServers can be moved out of register API and then all this logic can remain in callingClient

Let's discuss the approach over call if you need help

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 9]
```diff
@@ -433,39 +123,23 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
    * calls are cleaned up.
    */
   private callsClearedHandler = async () => {
-    if (!this.registration.isDeviceRegistered()) {
+    // this is a temporary logic to get registration obj
+    // it will change once we have proper lineId and multiple lines as well
+    const {registration} = Object.values(this.lineDict)[0];
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/CallingClient.ts:128]: If we do have multiple lines, what is the logic in that case? Why can't we use it now itself?

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 1 above)
Comment (Kesari3008) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:254]: A few fields here will get populated once we get a registration response from Mobius so we can remove those from here:

"lastSeen" : "2023-04-18T13:04:53Z",
"addresses" : [ "sip:ce4itauc6t_GM1ILJVUKPEC@[64941297.int10.bcld.webex.com

  "keepaliveInterval" : 30,
  "callKeepaliveInterval" : 900,
  "rehomingIntervalMin" : 60,
  "rehomingIntervalMax" : 120,
  "voicePortalNumber" : null,
  "voicePortalExtension" : null
}

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 10]
```diff
@@ -585,6 +234,44 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
       }
     );
   }
+
+  /**
+   *
+   */
+  private createLine() {
+    const lineId = uuid();
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/CallingClient.ts:242]: this action we can perform inside LIne class constructor

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 11]
```diff
@@ -585,6 +234,44 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
+      this.sdkConfig
+    );
+
+    this.lineDict[lineId] = lineInstance;
+  }
+
+  /**
+   *
+   */
+  public getLines() {
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/CallingClient.ts:272]: We maintain return types as well and mention them in the function definition. Refer to makeCall method.
We should do that for both these methods : createline() as well as getLines()

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 3 above)
Comment (Kesari3008) [reply packages/calling/calling/src/CallingClient/line/index.ts:119]: lineId, extension, phone number, voicemail are a few properties that will be coming from the config file from DMS but since we don't have that implemented for now, we can leave these values blank. 

Other values like intervals, lastSeen and sipAddresses, can be populated once we get registration response from Mobius.

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+import {METRIC_EVENT, REG_ACTION, METRIC_TYPE} from '../metrics/types';
+import {LOGGER, LogContext} from '../../Logger/types';
+import {handleRegistrationErrors, filterMobiusUris, validateServiceData} from '../../common';
+import SDKConnector from '../../SDKConnector';
+import {ICallManager} from '../calling/types';
+
+export default class Line extends Eventing<LineEventTypes> implements ILine {
+  #webex: WebexSDK;
+
+  #mutex: Mutex;
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/line/index.ts:44]: Why are we using # instead of saying private here? Any particular reason for the same?

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+    lastSeen: string,
+    keepaliveInterval: number,
+    callKeepaliveInterval: number,
+    rehomingIntervalMin: number,
+    rehomingIntervalMax: number,
+    voicePortalNumber: number,
+    voicePortalExtension: number,
+    callManager: ICallManager,
+    mutex: Mutex,
+    config?: CallingClientConfig
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/line/index.ts:106]: only SDK config is something relevant in this file so we can just pass that instead of passing whole callingClientConfig

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+  }
+
+  /**
+   * Wrapper to for device  deregister.
+   */
+  public async deregister() {
+    this.#registration.deregister();
+  }
+
+  // will be done as part of call related jira
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/line/index.ts:370]: We can remove these changes for now. Not required

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 15]
```diff
@@ -0,0 +1,463 @@
...
+    rehomingIntervalMin: number,
+    rehomingIntervalMax: number,
+    voicePortalNumber: number,
+    voicePortalExtension: number,
+    callManager: ICallManager,
+    mutex: Mutex,
+    config?: CallingClientConfig
+  ) {
+    super();
+    this.userId = userId;
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/line/index.ts:109]: We need one more field here in the line object : MobiusDeviceId. We are going to maintain lineId and MobiusDeviceId both.

**Review by Kesari3008 — CHANGES_REQUESTED** [review]

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 1 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts]: Removed the dummy values. Will refactor other fields coming from Mobius api response in a separate PR as discussed with Priya.

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts]: Done

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 3 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts:100]: Done

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 4 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts]: Moved getMobiusServers to calling client. As we can't call async method from constructor, created a init() method which will be called while creating calling client instance. Updated the logic in calling.js file as well but for a case where initializeClients() is called from constructor(when webex object is passed), need to update the code after discussing the correct approach.

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 5 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:284]: Done but please check above comment as well.

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 6 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:508]: Added description.

File: docs/samples/calling/app.js [hunk 16]
```diff
@@ -314,9 +314,11 @@ function userSession() {
 }
 
 function createDevice() {
-  callingClient.register();
+  const lineObj = Object.values(callingClient.getLines())[0];
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply docs/samples/calling/app.js]: Done

File: packages/calling/calling/package.json [hunk 17]
```diff
@@ -41,7 +41,7 @@
     "buffer": "6.0.3",
     "jest-html-reporters": "3.0.11",
     "platform": "1.3.6",
-    "uuid": "8.3.2",
+    "uuid": "^3.3.2",
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/package.json]: Moved it back to original version and updated code to correctly use it as per new version. Tested and it's working fine.

File: packages/calling/calling/src/CallingClient/line/types.ts [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+  dummy: any;
+}
+
+export interface ICallerInfo {
+  // will be done as part of call related jira
+  dummy: any;
+}
+
+export enum LineStatus {
+  INACTIVE = 'inactive',
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/types.ts:11]: Done

File: packages/calling/calling/src/CallingClient/line/types.ts [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+  REGISTERED = 'registered',
+  UNREGISTERED = 'unregistered',
+}
+
+export type LineEventTypes = {
+  [LINE_EVENTS.CONNECTING]: () => void;
+  [LINE_EVENTS.ERROR]: (error: LineError) => void;
+  [LINE_EVENTS.RECONNECTED]: () => void;
+  [LINE_EVENTS.RECONNECTING]: () => void;
+  [LINE_EVENTS.REGISTERED]: (deviceInfo: IDeviceInfo) => void;
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/types.ts]: Done

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 9 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:370]: As discussed with Priya, no need to refactor it as anyways this logic is going to change.

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 10 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts]: Done

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 11 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts]: Done

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 13 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts]: Renamed CallingClientConfig to SdkConfig as config object contains sdkConfig only.

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 14 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts]: Removed

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 15 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts:89]: Done

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+
+  public userId: string;
+
+  public clientDeviceUri: string;
+
+  public lineId: string;
+
+  public mobiusDeviceId: string | undefined;
+
+  public mobiusUri: string | undefined;
```
Comment (arun3528) [inline packages/calling/calling/src/CallingClient/line/index.ts:38]: @meta-dipanshu-sharma @Kesari3008  some of these values can be with _. if they are only used internally , for developer mobiusUri is not needed

File: packages/calling/calling/src/CallingClient/line/types.ts [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+export enum LINE_EVENTS {
+  CONNECTING = 'connecting',
+  ERROR = 'error',
+  RECONNECTED = 'reconnected',
+  RECONNECTING = 'reconnecting',
+  REGISTERED = 'registered',
+  UNREGISTERED = 'unregistered',
+}
+
+export interface ICall {
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/line/types.ts:24]: we shouldn't move this in this story

File: packages/calling/calling/src/CallingClient/line/types.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (Kesari3008) [reply packages/calling/calling/src/CallingClient/line/types.ts:69]: I still see IDeviceInfo being sent back. Change this

File: packages/calling/calling/src/CallingClient/metrics/index.ts [hunk 22]
```diff
@@ -262,10 +262,12 @@ class MetricManager implements IMetricManager {
  * @param webex - Webex object to communicate with metrics microservice.
  * @param indicator - Service Indicator.
  */
-export const getMetricManager = (webex: WebexSDK, indicator: ServiceIndicator): IMetricManager => {
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/metrics/index.ts:265]: Why was this change required

File: packages/calling/calling/src/CallingClient/registration/register.ts [hunk 23]
```diff
@@ -69,7 +70,8 @@ export class Registration implements IRegistration {
   private rehomingIntervalMax: number;
   private mutex: Mutex;
   private metricManager: IMetricManager;
-  private emitter: CallingClientEmitterCallback;
+  private emitter: LineEmitterCallback;
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/registration/register.ts:73]: call this lineEmitter now that we calliClient one as well

File: packages/calling/calling/src/CallingClient/registration/register.ts [hunk 24]
```diff
@@ -672,11 +646,11 @@ export class Registration implements IRegistration {
           body,
           (clientError, finalError) => {
             if (finalError) {
-              this.emitter(EVENT_KEYS.ERROR, undefined, clientError);
+              this.callingClientEmitter(EVENT_KEYS.ERROR, clientError);
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/registration/register.ts:649]: We should emit line error here, applicable at other places also in this file. From this file all the should be of type line error

File: packages/calling/calling/src/Errors/types.ts [hunk 25]
```diff
@@ -58,6 +59,10 @@ export type ErrorObject = {
   context: ErrorContext;
 };
 
+export interface LineErrorObject extends ErrorObject {
```
Comment (Kesari3008) [inline packages/calling/calling/src/Errors/types.ts:62]: Maybe we should also have a unique ID being sent as part of line error object

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 26]
```diff
@@ -0,0 +1,238 @@
...
+      device,
+      keepaliveInterval,
+      callKeepaliveInterval,
+      rehomingIntervalMin,
+      rehomingIntervalMax,
+      voicePortalNumber,
+      voicePortalExtension,
+    } = deviceInfo;
+
+    this.mobiusDeviceId = device?.deviceId;
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/line/index.ts:181]: Is the device object optional inside deviceInfo? Please check

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 27]
```diff
@@ -0,0 +1,238 @@
...
+    this.registration.deregister();
+  }
+
+  /**
+   * Request for the Device status.
+   *
+   * @param deviceInfo - Registered device Object.
+   * @returns Promise.
+   */
+  public async sendKeepAlive(deviceInfo: IDeviceInfo): Promise<void> {
```
Comment (Kesari3008) [inline packages/calling/calling/src/CallingClient/line/index.ts:146]: this API is going it be private, we can change it

**Review by Kesari3008 — CHANGES_REQUESTED** [review]

File: packages/calling/calling/src/CallingClient/CallingClient.ts (hunk 4 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts]: Done

File: packages/calling/calling/src/CallingClient/line/types.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/types.ts:69]: Priya, as we discussed in our last call that this will be deviceInfo only coming from mobius response and emitted from register.ts file. After we receive it in the line class in lineEmitter method, we will refactor the response and then will emit final object as ILine.

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 28]
```diff
@@ -225,6 +155,39 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
+    const line = Object.values(this.lineDict)[0];
+
+    setInterval(async () => {
+      if (
+        !this.webex.internal.mercury.connected &&
+        !retry &&
+        !Object.keys(this.callManager.getActiveCalls()).length
+      ) {
+        log.warn(`Network has flapped, waiting for mercury connection to be up`, {
+          file: REGISTRATION_FILE,
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:175]: Done

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 29]
```diff
@@ -261,15 +224,15 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
+          this.metricManager.submitRegistrationMetric(
             METRIC_EVENT.REGISTRATION_ERROR,
             REG_ACTION.REGISTER,
             METRIC_TYPE.BEHAVIORAL,
             clientError
           );
           this.emit(EVENT_KEYS.ERROR, clientError);
         },
-        {method: GET_MOBIUS_SERVERS_UTIL, file: CALLING_CLIENT_FILE}
+        {method: GET_MOBIUS_SERVERS_UTIL, file: LINE_FILE}
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:235]: Did the change locally earlier only but forgot to push. Pushed it now.

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 30]
```diff
@@ -585,11 +477,45 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
+      this.backupMobiusUris,
+      this.callingClientEmitter
+    );
+
+    this.lineDict[line.lineId] = line;
+  }
+
+  /**
+   * Retrieves details of all the line objects belonging to a User
+   * NOTE: currently multiple lines are not supported
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:502]: Done

File: packages/calling/calling/src/CallingClient/CallingClient.ts [hunk 31]
```diff
@@ -62,27 +62,27 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
...
 
   public mediaEngine: typeof Media;
 
+  private lineDict: Record<string, ILine> = {};
+
   /**
    * @param webex - A webex instance.
    * @param config - Config to start the CallingClient with.
    */
-  constructor(webex: WebexSDK, config?: CallingClientConfig) {
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/CallingClient.ts:85]: Changed it back to CallingClientConfig everywhere.

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 32]
```diff
@@ -0,0 +1,238 @@
...
+
+  public sipAddresses: string[] | undefined;
+
+  public voicemail: string | undefined;
+
+  public lastSeen: string | undefined;
+
+  public keepaliveInterval: number | undefined;
+
+  public callKeepaliveInterval: number | undefined;
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts:54]: Done

File: packages/calling/calling/src/CallingClient/line/types.ts [hunk 33]
```diff
@@ -0,0 +1,79 @@
+import {IRegistration} from '../registration/types';
+import {LineError} from '../../Errors/catalog/LineError';
+import {IDeviceInfo, MobiusDeviceId, MobiusStatus} from '../../common/types';
+
+export interface ICallerInfo {
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/types.ts:5]: Done

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 34]
```diff
@@ -0,0 +1,238 @@
...
+    );
+
+    if (this.registration.isDeviceRegistered()) {
+      this.registration.startKeepaliveTimer(url, interval);
+    } else {
+      log.warn('Device is not active, exiting.', logContext);
+    }
+  }
+
+  private fillLine(deviceInfo: IDeviceInfo) {
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts:170]: Done

File: packages/calling/calling/src/CallingClient/metrics/index.ts (hunk 22 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/metrics/index.ts:265]: fixed a eslint warning which says that every file should have a default export.

File: packages/calling/calling/src/CallingClient/registration/register.ts (hunk 23 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/registration/register.ts:73]: Done

File: packages/calling/calling/src/CallingClient/types.ts [hunk 35]
```diff
@@ -24,15 +17,14 @@ interface DiscoveryConfig {
   region: string;
 }
 
-export interface CallingClientConfig {
+export interface SdkConfig {
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/types.ts:20]: Done

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 26 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts:181]: Yes device is optional inside IDeviceInfo. Do we want to make it mandatory?

File: packages/calling/calling/src/CallingClient/registration/register.ts (hunk 24 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/registration/register.ts:649]: `handleRegistrationErrors` is called from both `register.ts` and `callingClient.ts`. We can't assign `LineErrorEmitterCallback` and `CallingClientErrorEmitterCallback` to handleRegistrationErrors emitterCb parameter. We can discuss more on this.

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 12 above)
Comment (arun3528) [reply packages/calling/calling/src/CallingClient/line/index.ts:22]: awesome thanks did not know that

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 26 above)
Comment (Kesari3008) [reply packages/calling/calling/src/CallingClient/line/index.ts:181]: No this is fine. I just wanted to check

File: packages/calling/calling/src/CallingClient/line/types.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (Kesari3008) [reply packages/calling/calling/src/CallingClient/line/types.ts:51]: That's correct but this event represents the event we emit from the line class so the event type should be ILine. Let's discuss this tomorrow if it is still not clear

File: packages/calling/calling/src/CallingClient/line/index.ts [hunk 36]
```diff
@@ -0,0 +1,238 @@
...
+
+  public userId: string;
+
+  public clientDeviceUri: string;
+
+  public lineId: string;
+
+  public mobiusDeviceId?: string;
+
+  public mobiusUri?: string;
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts:38]: Arun, as per this comment of Priya: https://github.com/webex/webex-js-sdk/pull/2995#discussion_r1298260273
we decided to keep it public

File: packages/calling/calling/src/CallingClient/registration/register.ts (hunk 24 above)
Comment (Kesari3008) [reply packages/calling/calling/src/CallingClient/registration/register.ts:649]: As discussed, this will be taken care in next story as it requires significant amount of changes

**Review by Kesari3008 — APPROVED** [review]
Looks good. Just want to discuss remaining open comments

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 13 above)
Comment (Kesari3008) [reply packages/calling/calling/src/CallingClient/line/index.ts]: Some confusion is there her, let's discuss this once

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 20 above)
Comment (arun3528) [reply packages/calling/calling/src/CallingClient/line/index.ts]: @Kesari3008  it might be needed for the web client for something else but not needed for developer to know correct ?

we could hide this based on type we expose if we don't want to expose to developers

**Review by arun3528 — APPROVED** [review]
@Kesari3008  lets have further discussion on the variable and types we expose

File: packages/calling/calling/src/CallingClient/line/types.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/types.ts:51]: Done

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 13 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts]: Done

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 20 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts]: As per discussion, made `mobiusUri` as private property.

File: packages/calling/calling/src/CallingClient/line/index.ts (hunk 36 above)
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/calling/src/CallingClient/line/index.ts:38]: Made it private
